### PR TITLE
Support for nested shadow DOM added

### PR DESCRIPTION
**Pull request by gabrielo91**
# Description

- New unit tests added to the functions  `isShadowElement`, `getShadowRoot`, `getShadowHost` and  `insertShadowSlots`
- Added support for insertShadowSlot function with nested web components

### CODE REVIEW

File: src/dom.js [hunk 1]
```diff
@@ -1095,7 +1095,7 @@ export function getShadowHost(element : Node) : ?HTMLElement {
     const shadowRoot = getShadowRoot(element);
 
     // $FlowFixMe
-    if (shadowRoot.host) {
+    if (shadowRoot && shadowRoot.host) {
```
Comment (gregjopa) [inline src/dom.js:1098]: Does this change allow us to remove the `// $FlowFixMe` comment above?

File: test/tests/dom/shadowDOMFunctions.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    if (!customElement || !customElement.shadowRoot) {
+        throw new Error('custom element does not have shadow root');
+    }
+
+});
+
+describe('isShadowElement cases', () => {
+  
+    it('should return true if parent node is shadow root',  () => {
+        const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:42]: Nice selector! I like the use of `?`

File: test/tests/dom/shadowDOMFunctions.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            console.log(error.message);
+            if (!error.message.match(/Element is not in shadow dom/i)) {
+                throw new Error(`should have thrown 'Element is not in shadow dom' exception, gotten: ${ error.message }`);
+            }
+        }
+
+    });
+
+    // don't know how to test it
+    it.skip('should throw exception if Host element is also in shadow dom', () => {
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:124]: I think we need another web component to be able to test this use case. Ex: `customElements.define('custom-web-component-wrapper' ...` Can you try this out?

File: test/tests/dom/shadowDOMFunctions.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+});
+
+describe('isShadowElement cases', () => {
+  
+    it('should return true if parent node is shadow root',  () => {
+        const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
+
+        if (!innerElement) {
+            throw new Error('there is not inner element');
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:45]:
```suggestion
            throw new Error('unable to find inner element');
```

Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js]: Fixed! thanks!

File: src/dom.js (hunk 1 above)
Comment (gabrielo91) [reply src/dom.js:1098]: actually no, even with that extra validation, that comment is needed

File: test/tests/dom/shadowDOMFunctions.js (hunk 3 above)
Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js]: Thanks for the suggestion! and just for record, it was no as trivial as doing this:

```html
<html>
      ...
     <wrapper-custom-component>
             <inner-component>
                    <div id="inner-div">
                          Text from a nested custom component
                    <div>
             </inner-component>
     </wrapper-custom-component>
    ...
</html>
```

What I did for testing the failing case (when shadoRoot is also shadowHost in `insertShadowSlot()` function) was to attach the  ` <inner-component>` to a new div designed as a host within the `<wrapper-custom-component>` and I could reproduce the expection`Host element is also in shadow dom`

File: test/tests/dom/shadowDOMFunctions.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            throw new Error('unable to find inner element');
+        }
+        const result = getShadowRoot(innerElement);
+
+        if (!result) {
+            throw new Error(`should have returned innerElement`);
+        }
+
+        if (!result.toString() === '[object ShadowRoot]') {
+            throw new Error(`should have returned '[object ShadowRoot]', gotten ${ result.toString() }`);
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:89]: Nit: for the error messages, could we replace all the "gotten" occurrences with "got"?
```suggestion
            throw new Error(`should have returned '[object ShadowRoot]', got ${ result.toString() }`);
```

Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js:89]: Sure! let me fix it

File: test/tests/dom/shadowDOMFunctions.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        const testSpan = document.createElement('span');
+        testSpan.setAttribute('id', 'inner-span');
+        testSpan.textContent = 'text from custom web component';
+        // Append it to the shadow root
+        shadowDOMContainer.appendChild(testSpan);
+        shadowHost.appendChild(shadowDOMContainer);
+    }
+};
+
+before(() => {
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:30]: It seems unusual to have a `before()` block outside of the `describe` blocks. With each test, we want to make sure we have a clean DOM. That way tests are isolated from one another and will not share any expected state in the DOM. To accomplish this, can we use a `beforeEach()` instead? Here's what I'm thinking:

```js

describe('web components', () => {
    describe('isShadowElement cases', () => {
        beforeEach(() => {
            // clean the DOM
            document.body.innerHTML = '';

            // do the setup required for the specific describe block we are in (ex: some blocks do not need custom-wrapper)
            customElements.define('custom-web-component', customWebComponent);
            customElements.define('custom-wrapper', customWebWrapper);

            if (!body) {
                throw new Error('Body not found');
            }

            const customElement = document.createElement('custom-web-component');
            body.appendChild(customElement);

            if (!customElement || !customElement.shadowRoot) {
                throw new Error('custom element does not have shadow root');
            }

        });

    describe('getShadowRoot cases', () => {

    describe('getShadowHost cases', () => {
})
```

Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js]: changed! thanks Greg!

File: test/tests/dom/shadowDOMFunctions.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+             *              #shadow-root (open)
+             *                <span id="inner-span"></span>
+             *           </div>
+             *    </custom-wrapper>
+             * </html>
+             */
+             
+    
+            try {
+                insertShadowSlot(innerSpan);
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:174]: I worry about false positives with this test. For example, if `insertShadowSlot()` does not throw an error, this test will still pass. What do you think about doing something like this so the test only passes when the error is thrown and passes the regex:

```js
let insertShadowSlotError;

try {
    insertShadowSlot(innerSpan);
} catch (error) {
    insertShadowSlotError = error;
}

if (!insertShadowSlotError?.error?.message.match(/Host element is also in shadow dom/)) {
    throw new Error(`should have thrown 'Host element is also in shadow dom' exception, got '${ error.message }'`);
}
```

Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js:175]: You´re  right, thanks for catching it, solved...

Comment (gregjopa) [reply test/tests/dom/shadowDOMFunctions.js:175]: Looks great! Can you use the same approach for the test above too? `should throw exception if element is not in shadow DOM`

Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js:178]: sure! fixed

**Review by gregjopa — APPROVED** [review]
Nice work! 💯

File: test/tests/dom/shadowDOMFunctions.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            }
+    
+            const result = insertShadowSlot(innerElement);
+    
+            if (!result) {
+                throw new Error('should have returned an element, got undefined');
+            }
+    
+            if (!result?.getAttribute('slot')?.match(/shadow-slot-/i)) {
+                throw new Error('should have returned an valid slot element');
```
Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:203]:
```suggestion
                throw new Error('should have returned a valid slot element');
```

Comment (gregjopa) [inline test/tests/dom/shadowDOMFunctions.js:203]:
```suggestion
                throw new Error('should have returned a valid slot element');
```

Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js:203]: fixed! thanks

File: src/dom.js [hunk 9]
```diff
@@ -1095,33 +1094,58 @@ export function getShadowHost(element : Node) : ?HTMLElement {
...
     const slot = document.createElement('slot');
     slot.setAttribute('name', slotName);
     element.appendChild(slot);
-    
+
     const slotProvider = document.createElement('div');
     slotProvider.setAttribute('slot', slotName);
     shadowHost.appendChild(slotProvider);
 
+    if (styles) {
```
Comment (bluepnume) [inline src/dom.js:1132]: Can you elaborate on why we need CSS rules here?

File: test/tests/dom/shadowDOMFunctions.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
+/* @flow */
+import { isShadowElement, getShadowRoot, getShadowHost, insertShadowSlot } from '../../../src';
```
Comment (bluepnume) [inline test/tests/dom/shadowDOMFunctions.js:2]: Tests look good to me at a glance!

**Review by bluepnume — APPROVED** [review]

File: src/dom.js [hunk 11]
```diff
@@ -1095,33 +1094,58 @@ export function getShadowHost(element : Node) : ?HTMLElement {
...
 
     if (!shadowHost) {
         throw new Error(`Element is not in shadow dom`);
     }
 
-    if (isShadowElement(shadowHost)) {
-        throw new Error(`Host element is also in shadow dom`);
-    }
+    // $FlowFixMe
+    let styleNode = getRootNode(element).querySelector('style');
```
[the remark this answers is not in the document]
Comment (gabrielo91) [reply src/dom.js:1121]: you´re right! thanks

File: src/dom.js [hunk 12]
```diff
@@ -1101,27 +1100,52 @@ export function getShadowHost(element : Node) : ?HTMLElement {
     }
 }
 
-export function insertShadowSlot(element : HTMLElement) : HTMLElement {
+  
+export function getRootNode(element : Node) : Node {
```
[the remark this answers is not in the document]
Comment (gabrielo91) [reply src/dom.js:1104]: good point, changed!

File: src/dom.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (gabrielo91) [reply src/dom.js:1121]: you're right, fixed, thanks!

File: test/tests/dom/shadowDOMFunctions.js [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    });
+    
+    describe('getShadowRoot cases', () => {
+        it('should return shadow root host', () => {
+            const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
+    
+            if (!innerElement) {
+                throw new Error('unable to find inner element');
+            }
+            const result = getShadowRoot(innerElement);
```
Comment (mnicpt) [inline test/tests/dom/shadowDOMFunctions.js:87]: @gabrielo91  The test says it should return shadow root host but we are just getting the root.  Is this correct?

**Review by mnicpt — APPROVED** [review]
Nice work, @gabrielo91 !  🎉

File: test/tests/dom/shadowDOMFunctions.js (hunk 13 above)
Comment (gabrielo91) [reply test/tests/dom/shadowDOMFunctions.js:87]: My mistake, the test declaration should be `should return shadow root`. Fixed

**Pull request merged**